google-cloud-bigquery/lib/google/cloud/bigquery/external.rb

```diff
@@ -0,0 +1,2200 @@
+      #                         external: { my_ext_table: csv_table }
+      #
+      #   data.each do |row|
+      #     puts row[:name]
+      #   end
+      #
+      module External
+        ##
+        # @private New External from URLs and format
+        def self.from_urls urls, format = nil
+          external_format = source_format_for urls, format
+          if external_format.nil?
+            fail ArgumentError, "Unable to determine external table format"
+          end
+          external_class = table_class_for external_format
+          external_class.new.tap do |e|
+            e.gapi.source_uris = Array(urls)
+            e.gapi.source_format = external_format
+          end
+        end
+
+        ##
+        # @private Determine source_format from inputs
+        def self.source_format_for urls, format
+          val = { "csv"                    => "CSV",
+                  "json"                   => "NEWLINE_DELIMITED_JSON",
+                  "newline_delimited_json" => "NEWLINE_DELIMITED_JSON",
+                  "sheets"                 => "GOOGLE_SHEETS",
+                  "google_sheets"          => "GOOGLE_SHEETS",
+                  "avro"                   => "AVRO",
+                  "datastore"              => "DATASTORE_BACKUP",
+                  "backup"                 => "DATASTORE_BACKUP",
+                  "datastore_backup"       => "DATASTORE_BACKUP",
+                  "bigtable"               => "BIGTABLE"
+                }[format.to_s.downcase]
+          return val unless val.nil?
+          Array(urls).each do |url|
+            return "CSV" if url.end_with? ".csv"
+            return "NEWLINE_DELIMITED_JSON" if url.end_with? ".json"
+            return "AVRO" if url.end_with? ".avro"
+            return "DATASTORE_BACKUP" if url.end_with? ".backup_info"
+            if url.start_with? "https://docs.google.com/spreadsheets/"
+              return "GOOGLE_SHEETS"
+            end
+            if url.start_with? "https://googleapis.com/bigtable/projects/"
+              return "BIGTABLE"
+            end
+          end
+          nil
+        end
+
+        ##
+        # @private Determine table class from source_format
+        def self.table_class_for format
+          case format
+          when "CSV"                    then External::CsvTable
+          when "NEWLINE_DELIMITED_JSON" then External::JsonTable
+          when "GOOGLE_SHEETS"          then External::SheetsTable
+          when "BIGTABLE"               then External::BigtableTable
+          else
+            # AVRO and DATASTORE_BACKUP
+            External::Table
+          end
+        end
+
+        ##
+        # # Table
+        #
+        # External::Table and its subclasses represents an external data source
+        # that can be queried from directly, even though the data is not stored
+        # in BigQuery. Instead of loading or streaming the data, this object
+        # references the external data source.
+        #
```

Comment:
It might be worth mentioning that AVRO and DATASTORE_BACKUP formats will be represented with objects of this top-level `Table` type.

Reply:
Added.